Sources/Purchasing/Purchases/PurchasesOrchestrator.swift

```diff
@@ -823,19 +830,27 @@ private extension PurchasesOrchestrator {
     /// - Parameter restored: whether the transaction state was `.restored` instead of `.purchased`.
     private func purchaseSource(
         for productIdentifier: String,
+        transaction: StoreTransaction,
         restored: Bool
     ) -> PurchaseSource {
         let initiationSource: ProductRequestData.InitiationSource = {
             // Having a purchase completed callback implies that the transation comes from an explicit call
             // to `purchase()` instead of a StoreKit transaction notification.
-            let hasPurchaseCallback = self.purchaseCompleteCallbacksByProductID.value.keys.contains(productIdentifier)
+            // As long as the transaction date is _after_ the completion block was added.
+            let completionBlockDate = self.purchaseCompleteCallbacksByProductID.value[productIdentifier]?.date
+
+            switch (completionBlockDate, restored) {
+            case let (.some(completionBlockDate), false):
+                if transaction.purchaseDate > completionBlockDate {
+                    return .purchase
+                } else {
+                    return .queue
+                }
 
-            switch (hasPurchaseCallback, restored) {
-            case (true, false): return .purchase
                 // Note that restores initiated through the SDK with `restorePurchases`
                 // won't use this method since those set the initiation source explicitly.
-            case (true, true): return .restore
-            case (false, _): return .queue
+            case (.some, true): return .restore
```

Comment:
kinda feels like we're losing a little bit of readability in the transition from `true` -> `some`

Reply:
That's the only way to both check for it being `nil` and extract the value from it though. Any suggestion?